language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -302,18 +316,62 @@ unfoldFixOf pat arg k = do
             , vArg
             ]
 
--- | Infer the type of a 'Builtin'.
-inferTypeOfBuiltinM
+-- ###########################################
+-- ## Dealing with applications of builtins ##
+-- ###########################################
+
+-- | Infer the type components of a 'Builtin'. The annotation argument
+-- is required for the error if the name isn't found
+inferTypeComponentsOfBuiltinM
     :: (GShow uni, GEq uni, DefaultUni <: uni)
-    => Builtin ann -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+    => ann -> BuiltinName -> TypeCheckM uni ann (TypeComponents uni)
 -- We have a weird corner case here: the type of a 'BuiltinName' can contain 'TypedBuiltinDyn', i.e.
 -- a static built-in name is allowed to depend on a dynamic built-in type which are not required
 -- to be normalized. For dynamic built-in names we store a map from them to their *normalized types*,
 -- with the normalization happening in this module, but what should we do for static built-in names?
 -- Right now we just renormalize the type of a static built-in name each time we encounter that name.
-inferTypeOfBuiltinM (BuiltinName    _   name) = normalizeType $ typeOfBuiltinName name
+inferTypeComponentsOfBuiltinM ann (StaticBuiltinName name) = getStaticBuiltinTypeComponentsM ann name
 -- TODO: inline this definition once we have only dynamic built-in names.
-inferTypeOfBuiltinM (DynBuiltinName ann name) = lookupDynamicBuiltinNameM ann name
+inferTypeComponentsOfBuiltinM ann (DynBuiltinName name)    = lookupDynamicBuiltinTypeComponentsM ann name
+
+
+-- | Work along the type parameters of a builtin application, building a
+-- mapping from type parameters to the types they're being instantiated at.
+checkBuiltinTypeArgs
+    :: ann
+    -> BuiltinName
+    -> [TyName]                -- type parameters
+    -> [Type TyName uni ann]   -- type instantations
+    -> TypeCheckM uni ann (Norm.TypeMapping TyName uni ())
+checkBuiltinTypeArgs ann bn tyNames tyArgs =
+    case zipExact tyNames tyArgs of
+      Left LeftLonger  -> throwError $ BuiltinTypeErrorE ann (BuiltinUnderInstantiated bn)
+      Left RightLonger -> throwError $ BuiltinTypeErrorE ann (BuiltinOverInstantiated  bn)
+      Right namesAndTypes ->  do
+        mapM_ (mapM_ (\tyarg -> checkKindM (typeAnn tyarg) tyarg (Type ()))) namesAndTypes
```

Comment:
And I'd probably use `for` here, as below.

Reply:
> `mapM` is from before `Traversable` was in base, and has a needlessly strong `Monad` constraint

I agree, but then everybody uses `sequence`, even though by that logic we should use `sequenceA`.

Reply:
I agree that `sequenceA` is better :) I mostly find I just use `traverse`, though.

Reply:
> Is... this using the traversable instance of `(,)`? I'd probably just not do that. How about:
> 
> ```
> traverse_ (\(_, tyarg) -> checkKindM ...) namesAndTypes
> ```

That was my first attempt (but with `mapM_` instead of `traverse`), but I thought I'd be told off for being too simplistic! It's quite a common pattern (which it took me a while to get used to) in the codebase to use `fmap` relying on the `Functor` instances of things on things like `(,)` and `Either`  (and indeed `Term`).  This seemed like a natural extension of that. I'lll go back to the explicit version though, since it's clearer what's going on.

Reply:
At least in my head, tuples are usually "data-like" things, and the fact that they have `Functor/Applicative/Monad` instances is kind of a quirk. I'm willing to fight for never using them (although no doubt I'm a hypocrite).

`Either` is a bit different, since it is actually widely used as a short-circuiting monad.